staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,895 @@
+	netvifc1a1 "k8s.io/examples/staging/kos/pkg/client/clientset/versioned/typed/network/v1alpha1"
+	kosinformers "k8s.io/examples/staging/kos/pkg/client/informers/externalversions"
+	kosinternalifcs "k8s.io/examples/staging/kos/pkg/client/informers/externalversions/internalinterfaces"
+	kosinformersv1a1 "k8s.io/examples/staging/kos/pkg/client/informers/externalversions/network/v1alpha1"
+	koslisterv1a1 "k8s.io/examples/staging/kos/pkg/client/listers/network/v1alpha1"
+	netfabric "k8s.io/examples/staging/kos/pkg/networkfabric"
+)
+
+const (
+	// K8s API server reosurce name for NetworkAttachments.
+	attResourceName = "networkattachments"
+
+	// NetworkAttachments in network.example.com/v1alpha1
+	// fields names. Used to build field selectors.
+	attNodeFieldName   = "spec.node"
+	attVNIFieldName    = "spec.vni"
+	attHostIPFieldName = "status.hostIP"
+	attIPFieldName     = "status.ipv4"
+
+	// fields selector comparison operators.
+	// Used to build fields selectors.
+	equal    = "="
+	notEqual = "!="
+
+	// resync period for Informers caches. Set
+	// to 0 because we don't want resyncs.
+	resyncPeriod = 0
+)
+
+// vnState stores all the state needed for a Virtual Network for
+// which there is at least one NetworkAttachment local to this node.
+type vnState struct {
+	// mutex is used to access the map localAtts
+	// and remoteAttsInformer when starting it.
+	mutex sync.RWMutex
+
+	// remoteAttsInformer is an informer on the NetworkAttachments that are
+	// both: (1) in the Virtual Network the vnState represents, (2) not on
+	// this node. It is stopped when the last local NetworkAttachment in the
+	// Virtual Network associated with the vnState instance is deleted. To
+	// stop it, remoteAttsInformerStopCh must be closed.
+	remoteAttsInformer       k8scache.SharedInformer
+	remoteAttsInformerStopCh chan struct{}
+
+	// remoteAttsLister is a lister on the NetworkAttachments that are
+	// both: (1) in the Virtual Network the vnState represents, (2) not
+	// on this node. Since a Virtual Network cannot span multiple k8s API
+	// namespaces, it's a NamespaceLister.
+	remoteAttsLister koslisterv1a1.NetworkAttachmentNamespaceLister
+
+	// localAtts stores the namespaced names the local NetworkAttachments in the Virtual
+	// Network the vnState represents. A local attachment is added to it when
+	// it is seen for the first time, and removed from it when it is deleted
+	// from the local attachments informer cache. Upon being removed, a check
+	// on whether the attachment is the last one in localAtts is made, and if that's
+	// the case remoteAttsInformer is stopped by closing remoteAttsInformerStopCh,
+	// and vnState is removed from the map storing all the vnStates localAtts is used
+	// as a set where the elements are the keys, the map values can be ignored.
+	// TODO we should implement and use a set rather than a map for this
+	localAtts map[k8stypes.NamespacedName]interface{}
```

Comment:
The most efficient way to do this is to make it a map to `struct{}` --- one of those takes zero bits to represent.

Reply:
Implemented this improvement in 3f2a65b, thanks for the hint.